addons/project/models/project.py

```diff
@@ -932,15 +938,23 @@ def _send_rating_all(self):
     # Privacy
     # ---------------------------------------------------
 
-    def _change_privacy_visibility(self):
+    def _change_privacy_visibility(self, new_visibility):
         """
         Unsubscribe non-internal users from the project and tasks if the project privacy visibility
-        goes to a value different than 'portal'
+        goes from 'portal' to a different value.
+        If the privacy visibility is set to 'portal', subscribe back project and tasks partners.
         """
-        for project in self.filtered(lambda p: p.privacy_visibility != 'portal'):
-            portal_users = project.message_partner_ids.user_ids.filtered('share')
-            project.message_unsubscribe(partner_ids=portal_users.partner_id.ids)
-            project.mapped('tasks')._change_project_privacy_visibility()
+        for project in self:
+            if project.privacy_visibility == new_visibility:
+                continue
+            if new_visibility == 'portal':
+                project.message_subscribe(partner_ids=project.partner_id.ids)
+                for task in project.mapped('task_ids').filtered('partner_id'):
```

Comment:
Just to inform you, the mapped is not necessary on a relational field. And so it is equivalent to: 
```suggestion
                for task in project.task_ids.filtered('partner_id'):
```
🙂